src/servers/launcher/main.cpp

```diff
@@ -0,0 +1,849 @@
+    }
+    os.flags(flags);
+
+    return os;
+  }
+};
+
+struct Context {
+  const UUID* SERVER_ID;
+};
+
+static Context ctx{};
+
+class DBConnectionError: public std::runtime_error {
+  public:
+  DBConnectionError(const std::string& error) noexcept:
+    std::runtime_error(error) {}
+};
+
+class DB {
+  PGconn* c;
+
+  DB() :
+    c([]{
+      PGconn* c{PQconnectdb("")};
+      if (PQstatus(c) != CONNECTION_OK) {
+        throw DBConnectionError(PQerrorMessage(c));
+      }
+      return c;
+    }()) {}
+
+  public:
+  static std::expected<DB, std::string> tryCreate() noexcept {
+    try {
+      return DB();
+    } catch (const DBConnectionError& exc) {
+      return std::unexpected(exc.what());
+    }
+  }
+  static const DB create() noexcept {
+    while (true) {
+      auto maybeDB{DB::tryCreate()};
+      if (maybeDB.has_value()) {
+        return std::move(maybeDB).value();
+      } else {
+        std::cout << maybeDB.error() << std::endl;
+        std::this_thread::sleep_for(RETRY_DELAY);
+      }
+    }
+  }
+  DB(const DB&) = delete;
+  DB(DB&& other) noexcept: c(std::exchange(other.c, nullptr)) {}
+  DB& operator=(const DB&) = delete;
+  DB& operator=(DB&& other) noexcept {
+    std::swap(c, other.c);
+
+    return *this;
+  }
+  ~DB() { PQfinish(c); }
+
+  const PGconn* getConn() const noexcept { return c; }
+};
+
+class QueryError: public std::runtime_error {
+  public:
+  QueryError(const std::string& error) noexcept:
+    std::runtime_error(error) {}
+};
+
+class Query {
+  PGresult* r;
+  Query(const PGconn* conn, const std::string& sql) :
+    r([&conn, &sql]{
+      PGresult* r{PQexecParams(
+        const_cast<PGconn*>(conn),
+        sql.c_str(),
+        0,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        1
+      )};
+      {
+        const ExecStatusType s{PQresultStatus(r)};
+        if (s != PGRES_TUPLES_OK && s != PGRES_COMMAND_OK) {
+          throw QueryError(PQerrorMessage(conn));
+        }
+      }
+
+      return r;
+    }()) {}
+
+  public:
+  static std::expected<Query, std::string>
+  tryCreate(PGconn* conn, const std::string& sql) noexcept {
+    try {
+      return Query(conn, sql);
+    } catch (const QueryError& exc) {
+      return std::unexpected(exc.what());
+    }
+  }
+  // No `.create()` because it's expected that trying the same query a
+  // second time will have similar results.
+  Query(const Query&) = delete;
+  Query(Query&& other) noexcept: r(std::exchange(other.r, nullptr)) {}
+  Query& operator=(const Query&) = delete;
+  Query& operator=(Query&& other) noexcept {
+    std::swap(r, other.r);
+
+    return *this;
+  }
+  ~Query() { PQclear(r); }
+
+  void fetchScalar(auto& out) const noexcept { out = PQgetvalue(r, 0, 0); }
+};
+
+class LauncherRepository {
+  const DB& db;
+
+  public:
+  LauncherRepository(const DB& db) noexcept: db(db) {}
+
+  const std::expected<const UUID, std::string> tryRegister() const noexcept {
```

Comment:
Once again, the `noexcept` here is a lie. There are a ton of ways an exception might come up in this function.

Reply:
* OOM when constructing the string to say the `Query` construction failed;
* OOM when fetching the string from the stringstream; and
* OOM when building the string in the stringstream?

How many did I miss?